scripts/word_embeddings/tools/cooccur.cc

```diff
@@ -0,0 +1,369 @@
+    paths.emplace(file);
+  }
+
+  return std::make_tuple(paths, output, args);
+}
+
+// * Utils
+template <class T>
+class CircularBuffer : public ranges::view_facade<CircularBuffer<T>> {
+public:
+  CircularBuffer(size_t capacity) : _capacity(capacity) {
+    _data.reserve(capacity);
+  }
+
+  size_t capacity() const { return _capacity; }
+
+  size_t size() const { return _data.size(); }
+
+  void push_back(T item) {
+    if (size() < capacity()) {
+      _data.push_back(item);
+    } else {
+      _data.at(_pointer) = item;
+      _pointer = (_pointer + 1) % _capacity;
+    }
+  }
+
+  void clear() {
+    _data.clear();
+    _pointer = 0;
+  }
+
+private:
+  const size_t _capacity;
+  std::vector<T> _data;
+  size_t _pointer = 0;
+
+  // Ranges TS based iterator over circular buffer
+  friend ranges::range_access;
+  struct cursor {
+    cursor() = default;
+    cursor(size_t pointer, size_t counter, const CircularBuffer *parent)
+        : pointer(pointer), counter(counter), parent(parent) {}
+
+    T read() const { return parent->_data.at(pointer); }
+    bool equal(const cursor &other) const { return counter == other.counter; }
+    void next() {
+      // LIFO
+      if (pointer == 0) {
+        pointer = parent->size() - 1;
+      } else {
+        pointer -= 1;
+      }
+      counter++;
+    }
+
+    size_t pointer;
+    size_t counter;
+    const CircularBuffer *parent;
+  };
+  cursor begin_cursor() const {
+    if (size() == 0)
+      return cursor(_pointer, 0, this);
+    else if (_pointer == 0)
+      return cursor(size() - 1, 0, this);
+    else
+      return cursor(_pointer - 1, 0, this);
+  }
+  cursor end_cursor() const { return cursor(_pointer, size(), this); }
+};
+
+// * Input
+auto ReadVocab() {
+  std::string word;
+  std::string count;
+  int rank{0};
+  Vocab vocab;
+  while (std::cin >> word) {
+    std::cin >> count;
+    vocab[word] = {rank, std::stoi(count)};
+    rank++;
+  }
+  return vocab;
+}
+// * Co-occurence matrix construction
+std::mutex paths_m;
+std::mutex matrices_m;
+
+void ReadMatrix(std::queue<fs::path> &paths, queue<Matrix> &matrices,
+                const Vocab &vocab, const Arguments &args, uint32_t seed) {
+  assert(seed > 0);
+  std::string line;
+  CircularBuffer<uint32_t> history(args.window_size);
+  std::unique_ptr<Matrix> m = std::make_unique<Matrix>();
+
+  // Prepare subsampling
+  std::random_device r;
+  std::default_random_engine random_engine(r());
+  std::uniform_real_distribution<float> uniform_dist(0, 1);
+  std::vector<double> idx_to_pdiscard;
+  if (args.subsample) {
+    auto counts = vocab | rsv::values | rsv::values;
+    double sum_counts = std::accumulate(counts.begin(), counts.end(), 0);
+    double t = 1E-4;
+    for (const double count : vocab | rsv::values | rsv::values) {
+      idx_to_pdiscard.push_back(1 - std::sqrt(t / (count / sum_counts)));
+    }
+  }
+
+  while (true) {
+    fs::path path;
+    {
+      std::scoped_lock lock(paths_m);
```

Comment:
std::scoped_lock is cpp17. Old compilers don't support this. What about just using std::lock_guard?